text/3829-de-rfc-unsized-locals.md

```diff
@@ -0,0 +1,146 @@
+
+# Demotivation
+[demotivation]: #demotivation
+
+The `unsized_locals` feature is simple to explain on the surface: local variables no longer have to be `Sized`.
+
+```rust
+#![feature(unsized_locals)]
+fn main() {
+    let x = *Box::from("hello, world!");
+}
+```
+
+This will dynamically allocate space on the stack for the string.
+C has a similar feature, [`alloca`] and variable length arrays (VLA), the latter having been made optional to implement in C11.
+
+## Lack of Proper Implementation
+
+This feature has never been properly implemented in rustc.
+The variable length array form proposed in the RFC still doesn't exist at all, only the unsized local variable does.
+It is implemented in the type checker and codegen, but lacking MIR semantics and therefore unimplemented in the compile time function evaluator.
+This is very significant, as MIR semantics govern how the feature should behave precisely in the first place.
+Without them, they cannot work in `const` and optimizations are likely broken around them.
+Because of this lack of implementation quality, the `unsized_locals` feature was already accepted for removal from rustc two years ago in [MCP 630].
+This removal hasn't yet been implemented.
+
+## Implicit Danger
+
+While lack of implementation quality is a sign of lack of interest for the feature, it is not the primary reason for this RFC,
+which is purely about the langauge design of the feature.
+
+The original RFC was very short, and especially short on motivation and rationale for the design.
+
+Dynamic stack allocation in general has a rather significant downside: it makes it easy to accidentally overflow the stack if you allocate a lot.
+Stacks are usually rather small (on the order of few megabytes, depending on the platform), which means that dynamically allocating user-controlled input on the stack is often rather dangerous.
+While stack overflows are not considered memory unsafe by Rust, they still cause crashes which can lead to denial of service vulnerabilities and unreliability in general.
+Dynamic stack allocation as currently implemented interacts especially poorly with loops.
+Allocations are not freed until the function returns, so the following example overflows the stack:
+
+```rust
+#![feature(unsized_locals)]
+fn main() {
+    let s = "A".repeat(1000);
+    for i in 0..1000000 {
+        let x: str = *Box::from(s.as_str());
+        std::hint::black_box(&s);
+    }
+}
+```
+There are ways around this (rewinding the stack pointer at the end of the loop to free up the memory), but they are not currently implemented.
+
+Dynamic stack allocation also has its upsides.
+It is generally faster than heap allocation and can therefore improve performance in cases where the previously mentioned downsides are not a concern.
+Therefore, this RFC is not necessarily a rejection of the idea of dynamic stack allocation, but merely the way the `unsized_locals` feature exposes it.
```

Comment:
I agree that we still need to figure out the best way to expose the feature, but I think that the best way to do that is to implement it properly on nightly and enable experimentation. Right now, we have vague concerns about footguns and blowing the stack, but without concrete user feedback, it’s very hard to know how severe the issue is or how the problem could be mitigated.

Reply:
E.g. maybe lints could address the danger. Or maybe all unsized locals should be annotated in some way (e.g. `let #[unsized] foo = …`). We can’t try those things out unless we have a proper implementation.

Reply:
Then go implement it if you think it's such a good idea.

Reply:
No significant experimentation has happened for the past **eight** years. I wouldn't hold my beer.

Reply:
Right, it's actually the middle-end, I think?

Reply:
> No significant experimentation has happened for the past **eight** years.

Yes, because the middle/back-end implementation is broken, scaring nightly users away. If the compiler team would rather strip out that broken implementation than continue to let it rot, I have no objection. But https://github.com/rust-lang/compiler-team/issues/630 is enough for that. IIUC this de-RFC would require anyone motivated to implement the feature properly, to go through the full RFC process first. I think that’s going too far. My preference would be to “downgrade” #1909 to an eRFC, instead of unaccepting it completely.

Reply:
Changing it to an "experimental" RFC will not change that right now it is a gravestone of a feature that no one wants to proceed on. Yes, probably someone *should* go through the RFC process again, because right now there's not even a draft of a proposal for how it may be implemented in rustc's semantic layers.

Reply:
>  it is a gravestone of a feature that no one wants to proceed on

I wouldn’t say that. For example, someone recently put in the effort to fix the longstanding issue with unsized locals being misaligned. Clearly there are *some* people still interested in this.

Reply:
I believe that what @tmiasko was touching on there involved fixes to parts of the code that were shared between `unsized_locals` and `unsized_fn_params`, because code that accepts `unsized_fn_params` sometimes, due to the way codegen works, creates unsized temporaries, i.e. locals. That then requires handling that for `unsized_fn_params`. Our current rude hack that addresses this involves blocking inlining for `unsized_fn_params`, since that's one of the conditions to reach that.

Also see:
- https://github.com/rust-lang/rust/issues/111355
- https://github.com/rust-lang/rust/issues/111175